### PR TITLE
MAILBOX-351 ReIndexing should be MailboxId based

### PR DESCRIPTION
**Pull request by chibenwa**
Impact:

 - Simplification of ReIdexerImpl internal (no need of MailboxPath rename tracking)
 - Enhance ReIndexer API: allow use of MailboxId
 - Improve reIndex routes webAdmin API: remove the useless user parameter where not needed.

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[1f359c96fabf164293ca04132ea5a88ab79bf39f] [ERROR]   FilterTest.messageShouldBeAppendedInInboxWhenFromDoesNotMatchRule:1081 » ConditionTimeout
[1f359c96fabf164293ca04132ea5a88ab79bf39f] [ERROR]   SendMDNMethodTest.sendMDNShouldSendAMDNBackToTheOriginalMessageAuthor:286->bartSendMessageToHomer:138 » ConditionTimeout
```

test this please

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 1]
```diff
@@ -70,16 +68,17 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
...
+        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
+        return reIndexSingleMailbox(mailbox.getMailboxId(), reprocessingContext);
     }
 
     Task.Result reIndex(ReprocessingContext reprocessingContext) throws MailboxException {
-        MailboxSession mailboxSession = mailboxManager.createSystemSession("re-indexing");
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(RE_INDEXING);
         LOGGER.info("Starting a full reindex");
-        List<MailboxPath> mailboxPaths = mailboxManager.list(mailboxSession);
+        List<Mailbox> mailboxes = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).list();
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: This looks suspicious to me that we require the mapper layer. A quick look at Jmap make me think that we should rather use `mailboxManager.search` here. WDYT ?

Comment (mbaechler) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: In term of design, we should refrain from using mapper almost everywhere, we will probably have to remove the mapper layer in the future.

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java [hunk 2]
```diff
@@ -157,6 +157,27 @@ void messageReIndexShouldBeWellPerformed() throws Exception {
             && message.getUid().equals(createdMessage.getUid()));
     }
 
+    @Test
+    void messageReIndexUsingMailboxIdShouldBeWellPerformed() throws Exception {
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java]: you probably need more than a single test for this new API.

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: To answer you, it seems really hard to set up this feature on top of `MailboxManager` for very good reasons:
 - API level: Indexers expects MailboxMessage, which is not part of mailbox API, but only defined at the mapper level. Thus we are forced at some point to use mappers.
 - Rights: We do not get yet "administrative" mailbox session (only impersonation). The bad thing with IDs is that you don't know the user, thus you don't know which username to be impersonating to. This is the issue removed in https://github.com/linagora/james-project/pull/1948/commits/1f359c96fabf164293ca04132ea5a88ab79bf39f

Also, the ReIndexer is only reading from the main store so calling it is safe - we won't encoutered the bugs faced 2.5 years ago with JMAP that an intern did spotted.

Finally, the overall idea of reIndexing is internal to the mailbox (indexing APIs are not exposed outside of this context) hence it does not shock me.

Anyway, acting on any of the two points I listed above is **clearly out of reach** of a 2 point ticket.

Anyway, I am going to split the PR in 2, but I still believe in what I explained here.

Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: Ok, got it: for this specific case we can still use `mailboxManager` as it was done before.

**Comment by chibenwa** [comment]
`MailQueueContract.concurrentEnqueueDequeueShouldNotFail:387 » NotTerminated`

Pff 

test this please

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java [hunk 3]
```diff
@@ -39,15 +45,28 @@
...
+        this.mailboxManager = mailboxManager;
+        this.mapperFactory = mapperFactory;
     }
 
     @Override
-    public Task reIndex(MailboxPath path) {
-        return new SingleMailboxReindexingTask(reIndexerPerformer, path);
+    public Task reIndex(MailboxPath path) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("ReIndexingImap");
+        Mailbox mailbox = mapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: mailboxManager.search ?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java [hunk 4]
```diff
@@ -39,15 +45,28 @@
...
+        return new SingleMailboxReindexingTask(reIndexerPerformer, mailbox);
+    }
+
+    @Override
+    public Task reIndex(MailboxId mailboxId) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("ReIndexingImap");
+        Mailbox mailbox = mapperFactory.getMailboxMapper(mailboxSession).findMailboxById(mailboxId);
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: mailboxManager.search ?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java [hunk 5]
```diff
@@ -61,7 +80,16 @@ public Task reIndex(User user) {
     }
 
     @Override
-    public Task reIndex(MailboxPath path, MessageUid uid) {
-        return new SingleMessageReindexingTask(reIndexerPerformer, path, uid);
+    public Task reIndex(MailboxPath path, MessageUid uid) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("ReIndexingImap");
+        Mailbox mailbox = mapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: mailboxManager.search ?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java [hunk 6]
```diff
@@ -61,7 +80,16 @@ public Task reIndex(User user) {
...
+        return new SingleMessageReindexingTask(reIndexerPerformer, mailbox, uid);
+    }
+
+    @Override
+    public Task reIndex(MailboxId mailboxId, MessageUid uid) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("ReIndexingImap");
+        Mailbox mailbox = mapperFactory.getMailboxMapper(mailboxSession).findMailboxById(mailboxId);
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: mailboxManager.search?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTask.java [hunk 7]
```diff
@@ -55,22 +59,22 @@ public int getFailedReprocessedMailCount() {
...
 
     private final ReIndexerPerformer reIndexerPerformer;
-    private final MailboxPath path;
+    private final Mailbox mailbox;
     private final AdditionalInformation additionalInformation;
     private final ReprocessingContext reprocessingContext;
 
     @Inject
-    public SingleMailboxReindexingTask(ReIndexerPerformer reIndexerPerformer, MailboxPath path) {
+    public SingleMailboxReindexingTask(ReIndexerPerformer reIndexerPerformer, Mailbox mailbox) {
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTask.java]: isn't MailboxId supposed to be enough ?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: Not possible: right would be denied as you don't know yet what user that mailbox does belong to.

We miss an always allowed session...

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java (hunk 6 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: Not possible: right would be denied as you don't know yet what user that mailbox does belong to.

We miss an always allowed session...

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTask.java (hunk 7 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTask.java]: I did not wanted to duplicate code between `mailboxPath` & `MailboxId`. This prouved to be convenient.

**Comment by chibenwa** [comment]
To answer this entire thread:

 - I want to rely on a common representation of a Mailbox, allowing to unify `MailboxPath` and `MailboxId` APIs.

After all, we want to reIndex a mailbox and we don't really care how the mailbox was designated once it is retrieved.

To represent a mailbox we have a couple of models choice:

 - Use a `MessageManager` -> this can be retrieved via `mailboxManager.getMailbox(XXX)` XXX being id or path. However right is enforced on the ID version (not on the path version ?). Thus as such, as we can not bypass or trick the right system (we don't know who to impersonate to as we get JUST a mailboxId).

**Comment by hoangdat** [comment]
Build triggered. sha1 is merged.

**Comment by hoangdat** [comment]
Build started sha1 is merged.

**Comment by hoangdat** [comment]
Build finished. No test results found.

**Comment by hoangdat** [comment]
Build triggered. sha1 is merged.

**Comment by hoangdat** [comment]
Build started sha1 is merged.

**Comment by chibenwa** [comment]
```
[99d6109c3c2c95b4cb880f1f758700acf8486de0] [ERROR]   SetMessagesMethodTest.attachmentsShouldBeRetrievedWhenChainingSetMessagesAndGetMessagesTextAttachment Multiple Failures (2 failures)
[99d6109c3c2c95b4cb880f1f758700acf8486de0] 	Failed to parse the JSON document
[99d6109c3c2c95b4cb880f1f758700acf8486de0] 	com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /192.168.0.1:57719 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/192.168.0.1:57719] Timed out waiting for server response))
```

test this please

**Comment by chibenwa** [comment]
`SwiftTempAuthObjectStorageBlobsDAOBuilderTest.builtBlobsDAOCanStoreAndRetrieve:94 » Timeout`

test this please

**Comment by chibenwa** [comment]
```
[99d6109c3c2c95b4cb880f1f758700acf8486de0] [ERROR]   SetMessagesMethodTest.setMessagesShouldTriggerMaxQuotaReachedWhenTryingToSendMessageAndQuotaReached Multiple Failures (2 failures)
[99d6109c3c2c95b4cb880f1f758700acf8486de0] 	All host(s) tried for query failed (tried: /192.168.0.1:57279 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/192.168.0.1:57279] Timed out waiting for server response))
[99d6109c3c2c95b4cb880f1f758700acf8486de0] 	com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /192.168.0.1:57279 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/192.168.0.1:57279] Timed out waiting for server response))
```

test this please

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 8]
```diff
@@ -66,10 +65,23 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
...
-        MailboxSession mailboxSession = mailboxManager.createSystemSession(path.getUser());
-        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
-        return reIndexSingleMailbox(mailbox.getMailboxId(), reprocessingContext);
+    Task.Result reIndex(MailboxId mailboxId, ReprocessingContext reprocessingContext) throws MailboxException {
+        LOGGER.info("Intend to reindex {}", mailboxId);
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(RE_INDEXING);
+        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxById(mailboxId);
+        messageSearchIndex.deleteAll(mailboxSession, mailbox);
+        try {
+            return Iterators.toStream(
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: could you extract the first 4 lines into a method `listMessageUidsInMailbox` ?

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java [hunk 9]
```diff
@@ -154,4 +159,89 @@ void messageReIndexShouldBeWellPerformed() throws Exception {
...
+
+        reIndexer.reIndex(mailboxId, createdMessage.getUid()).run();
+        ArgumentCaptor<MailboxMessage> messageCaptor = ArgumentCaptor.forClass(MailboxMessage.class);
+        ArgumentCaptor<Mailbox> mailboxCaptor = ArgumentCaptor.forClass(Mailbox.class);
+
+        verify(messageSearchIndex).add(any(MailboxSession.class), mailboxCaptor.capture(), messageCaptor.capture());
+        verifyNoMoreInteractions(messageSearchIndex);
+
+        assertThat(mailboxCaptor.getValue()).matches(mailbox -> mailbox.getMailboxId().equals(mailboxId));
+        assertThat(messageCaptor.getValue()).matches(message -> message.getMailboxId().equals(mailboxId)
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java]: please use assertJ assertion and not java equals

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java [hunk 10]
```diff
@@ -154,4 +159,89 @@ void messageReIndexShouldBeWellPerformed() throws Exception {
         assertThat(messageCaptor.getValue()).matches(message -> message.getMailboxId().equals(mailboxId)
             && message.getUid().equals(createdMessage.getUid()));
     }
+
+    @Test
+    void messageReIndexUsingMailboxIdShouldBeWellPerformed() throws Exception {
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java]: I think that you can replace "WellPerformed" with what you actually asserts : `shouldCallMessageSearchIndex`

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java [hunk 11]
```diff
@@ -154,4 +159,89 @@ void messageReIndexShouldBeWellPerformed() throws Exception {
...
+        verify(messageSearchIndex).add(any(MailboxSession.class), mailboxCaptor.capture(), messageCaptor.capture());
+        verifyNoMoreInteractions(messageSearchIndex);
+
+        assertThat(mailboxCaptor.getValue()).matches(mailbox -> mailbox.getMailboxId().equals(mailboxId));
+        assertThat(messageCaptor.getValue()).matches(message -> message.getMailboxId().equals(mailboxId)
+            && message.getUid().equals(createdMessage.getUid()));
+    }
+
+    @Test
+    void mailboxIdReIndexShouldBeWellPerformedWhenEmptyMailbox() throws Exception {
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java]: well performed here looks like a call to deleteAll and nothing more, right ?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java [hunk 12]
```diff
@@ -223,14 +221,25 @@ private TaskIdDto reIndexMailbox(Request request, Response response) {
...
-        Task task = taskGenerator.get();
-        TaskId taskId = taskManager.submit(task);
-        return TaskIdDto.respond(response, taskId);
+        try {
+            Task task = taskGenerator.generate();
+            TaskId taskId = taskManager.submit(task);
+            return TaskIdDto.respond(response, taskId);
+        } catch (MailboxNotFoundException e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java]: I don't think it's "bad request" but rather "not found" right ?

File: src/site/markdown/server/manage-webadmin.md [hunk 13]
```diff
@@ -2204,10 +2204,10 @@ concurrent changes done during the reIndexing might be ignored.
 ### ReIndexing a mailbox mails
 
 ```
-curl -XPOST http://ip:port/mailboxIndex/users/bob@domain.com/mailboxes/{mailboxId}?task=reIndex
+curl -XPOST http://ip:port/mailboxIndex/mailboxes/{mailboxId}?task=reIndex
```
Comment (mbaechler) [inline src/site/markdown/server/manage-webadmin.md]: I'm still convinced that mailboxIndex should not be a visible resource outside of James. Why don't we implement this action on `/mailboxes/` ?

Comment (mbaechler) [reply src/site/markdown/server/manage-webadmin.md]: * all mailboxes is easy, you omit the mailboxid and that's all
* per user is more tricky, you can pass it as a query string (filters are usually query params in REST)

Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: Maybe such API changes could be handled in a separate PR?

**Comment by chibenwa** [comment]
`MailQueueContract.concurrentEnqueueDequeueShouldNotFail:387 » NotTerminated`

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**